File: javalin/src/main/java/io/javalin/core/event/EventManager.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.core.event;

import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.function.Consumer;

public class EventManager {
    private final EnumMap<JavalinEvent, Set<EventHandler>> lifecycleHandlers = new EnumMap<>(JavalinEvent.class);

    private final Set<Consumer<HandlerMetaInfo>> handlerAddedHandlers = new LinkedHashSet<>();
    private final Set<Consumer<WsHandlerMetaInfo>> wsHandlerAddedHandlers = new LinkedHashSet<>();

    public EventManager() {
        for (JavalinEvent eventType : JavalinEvent.values()) {
            lifecycleHandlers.put(eventType, new HashSet<>());
        }
    }

    public void fireEvent(@NotNull JavalinEvent javalinEvent) {
        for (EventHandler eventHandler : lifecycleHandlers.get(javalinEvent)) {
            try {
                eventHandler.handleEvent();
            } catch (Exception e) {
                //TODO: Improve here - review if handleEvent() should throw?
                throw new RuntimeException(e);
            }
        }
    }

    public void fireHandlerAddedEvent(@NotNull HandlerMetaInfo metaInfo) {
        for (Consumer<HandlerMetaInfo> it : handlerAddedHandlers) {
            it.accept(metaInfo);
        }
    }
    public void fireWsHandlerAddedEvent(@NotNull WsHandlerMetaInfo metaInfo) {
        for (Consumer<WsHandlerMetaInfo> it : wsHandlerAddedHandlers) {
            it.accept(metaInfo);
        }
    }

    public EnumMap<JavalinEvent, Set<EventHandler>> getLifecycleHandlers() {
        return lifecycleHandlers;
    }

    public Set<Consumer<HandlerMetaInfo>> getHandlerAddedHandlers() {
        return handlerAddedHandlers;
    }

    public Set<Consumer<WsHandlerMetaInfo>> getWsHandlerAddedHandlers() {
        return wsHandlerAddedHandlers;
    }
}

File: src/main/java/io/javalin/core/event/EventManager.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.core.event

import io.javalin.core.security.Role
import io.javalin.http.HandlerType
import io.javalin.websocket.WsHandlerType
import java.util.*
import java.util.function.Consumer

class EventManager {
    val lifecycleHandlers = JavalinEvent.values().associate { it to HashSet<EventHandler>() }
    var handlerAddedHandlers = mutableSetOf<Consumer<HandlerMetaInfo>>()
    val wsHandlerAddedHandlers = mutableSetOf<Consumer<WsHandlerMetaInfo>>()
    fun fireEvent(javalinEvent: JavalinEvent) = lifecycleHandlers[javalinEvent]?.forEach { eventHandler -> eventHandler.handleEvent() }
    fun fireHandlerAddedEvent(metaInfo: HandlerMetaInfo) = handlerAddedHandlers.apply { this.forEach { it.accept(metaInfo) } }
    fun fireWsHandlerAddedEvent(metaInfo: WsHandlerMetaInfo) = wsHandlerAddedHandlers.apply { this.forEach { it.accept(metaInfo) } }
}

enum class JavalinEvent { SERVER_STARTING, SERVER_STARTED, SERVER_START_FAILED, SERVER_STOPPING, SERVER_STOPPED }
data class HandlerMetaInfo(val httpMethod: HandlerType, val path: String, val handler: Any, val roles: Set<Role>)
data class WsHandlerMetaInfo(val handlerType: WsHandlerType, val path: String, val handler: Any, val roles: Set<Role>)
